Tokenized PRBs on shortcode cart and checkout

mdmoore opened the pull request:
Fixes #8528

#### Changes proposed in this Pull Request

This implements the tokenized cart PRB on shortcode cart and checkout pages.
This is behind the `_wcpay_feature_tokenized_cart_prb` feature flag. Enable it in your `wp_options`.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

* Enable the "Enable Cart-Token implementation for PRBs" flag in your dev tools (you'll need to update them)
* Ensure you have Google Pay/Apple Pay enabled in the merchant's settings
* Go to a product page on your site
* You should be able to use the PRB button to get shipping data & check out on simple products & variable products
* You should be able to add a product to the cart and complete checkout from the shortcode cart and checkout pages

Borrowed from the [initial PDP implementation](https://github.com/Automattic/woocommerce-payments/pull/8644) 
* There shouldn't be a difference in functionality with the older implementation (when the "Enable Cart-Token implementation for PRBs" functionality is disabled) besides what's written below - if you do notice something weird, please let me know!
  * The implementation allows you to check out with the US shipping address, but for some reason the UK and CA addresses aren't working correctly (they fail validation in the backend, despite the values being correct)
  * The itemized amount (tax, shipping, etc.) is _sometimes_ not displayed

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

frosso committed:
feat: tokenized cart PRBs on PDPs

frosso committed:
WIP variations & compatibility work

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
better transformers organization

frosso committed:
WIP

frosso committed:
fix variable product page load

frosso committed:
revert some unrelated changes

frosso committed:
revert some unrelated changes

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
tests

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
PR review feedback

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
JSDocs

frosso committed:
typos

frosso committed:
WooPayments <> WcPay

frosso committed:
boolean prefix

frosso committed:
better variable naming

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
order attribution values

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
save point

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
adding await, just because

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
adding additional header to Store API response

frosso committed:
do not enqueue woopay checkout scripts on product pages

frosso committed:
Merge branch 'develop' into refactor/pdp-payment-request-tokenized-cart

frosso committed:
remove payment_request_api TODO

mdmoore committed:
WIP - Cart

mdmoore committed:
Merge 'develop'

mdmoore committed:
Init payment request on checkout

mdmoore committed:
Re-init PRB on cart total update and limit quantity change events to product page

mdmoore committed:
Update tests

mdmoore committed:
changelog

mdmoore committed:
Update test

mdmoore committed:
Update comment

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8891` or branch name `prb-tokenized-cart-shortcode-cart-checkout` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=prb-tokenized-cart-shortcode-cart-checkout) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 121c367f8e85f72141b091e02b232c7fa56c927c
- Build time: 2024-06-20 00:48:53 UTC

_Note: the build is updated when a new commit is pushed to this PR._

frosso commented on client/tokenized-payment-request/cart-api.js:
```diff
@@ -79,10 +80,17 @@ export default class PaymentRequestCartApi {
 		} );
+
+		this.cartRequestHeaders = {
+			Nonce: response.headers.get( 'Nonce' ),
```
What was the reason for copying the `Nonce` header, in this case?

The `Nonce` field retrieved from the request should also be copied in the `X-WooPayments-Store-Api-Nonce` property of `cartRequestHeaders` - similar to this: https://github.com/Automattic/woocommerce-payments/blob/b81abb4a3426c2bd47cc9a3cf5533e53cd7f5287/client/tokenized-payment-request/cart-api.js#L107-L107

FangedParakeet replied:
I think the only reason to retrieve and update the Nonce is if the nonce is not already present, when `createAnonymousCart()` has not been triggered before `getCart()`. AFAICT, the only way I can see this occurring is if we have `cachedCartData` and the nonce has been refreshed. Is this a feasible scenario or did you discover any other circumstance on the shortcode checkout that required us to fetch the nonce anew, @mdmoore?

mdmoore replied:
`Nonce` was copied here because it's needed for [the `updateCustomer()` call](https://github.com/Automattic/woocommerce-payments/blob/b81abb4a3426c2bd47cc9a3cf5533e53cd7f5287/client/tokenized-payment-request/payment-request.js#L235) that happens after clicking the the PRB. Without the nonce, that request fails with a nonce error. 

I see that `X-WooPayments-Store-Api-Nonce` will also work, since it overrides the `Nonce` when present, but I'm not sure it's needed with `getCart()` as `Nonce` alone works fine in block based themes. @frosso Can you refresh my memory on which point in the PRB flow the nonce error was occurring with block based themes that prompted the need to override it? If it's happening on the cart page as well, we can just use `X-WooPayments-Store-Api-Nonce` unless you think we should just use it regardless?

>AFAICT, the only way I can see this occurring is if we have cachedCartData and the nonce has been refreshed. Is this a feasible scenario or did you discover any other circumstance on the shortcode checkout that required us to fetch the nonce anew, @mdmoore?

On the cart page, a nonce has not yet been set in `cartRequestHeaders` by any previous request. The PDP uses `createAnonymousCart()` to add the nonce values to `cartRequestHeaders` but the cart page doesn't use the anonymous cart request and instead uses `getCart()`. So in short, whether adding a product to the cart on the shop page or from within the PDP, the nonce still must be copied to the headers so that it's present for subsequent Store API requests.

frosso replied:
I think I understand the problem a bit better.
The thing with the `Nonce` is that it's needed for `POST`/`PUT`/`DELETE` requests. It's not needed for `GET` requests.
There is a scenario where the `Nonce` is automatically saved, added, and updated to each request to the Store API: when [this middleware](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/middleware/store-api-nonce.js) is present on the page.
The middleware is always present when a block-based theme is used, or when a cart or checkout block is present on the page.
Since this implementation is for the classic checkout, the middleware is not present on the page. 
[This PR](https://github.com/Automattic/woocommerce-payments/pull/8840) has a bit more background on this issue.

Using the `Nonce` from the `getCart` response seems to be a good solution to me 👍 

It's unfortunate that in this scenario, the `Nonce` is not automagically provided by the middleware (because the middleware is not loaded).

mdmoore replied:
Ah, that makes sense as to why I wasn't seeing the nonce added automatically - no magic on the shortcode pages. Thanks for the explanation!

frosso commented on client/tokenized-payment-request/cart-api.js:
```diff
@@ -117,13 +125,14 @@ export default class PaymentRequestCartApi {
 			method: 'POST',
 			path: '/wc/store/v1/cart/update-customer',
-			credentials: 'omit',
+			credentials: context === 'product' ? 'omit' : 'same-origin',
```
Rather than passing an additional argument to these methods, what do you think about using the `getPaymentRequestData` utility, which is already used in this file?

```suggestion
			credentials: getPaymentRequestData( 'button_context' ) === 'product' ? 'omit' : undefined,
```

mdmoore replied:
Good call! [e8f655a](https://github.com/Automattic/woocommerce-payments/pull/8891/commits/e8f655a866e3aff1c19d9a1d952c21ee68a0d30f)

FangedParakeet commented on client/tokenized-payment-request/payment-request.js:
```diff
@@ -195,30 +198,35 @@ export default class WooPaymentsPaymentRequest {
+						window.alert(
 							__(
-								'Sorry, this product is unavailable. Please choose a different combination.',
+								'Please select your product options before proceeding.',
```
I'm not necessarily sure why we need to lean on WC Core's default messaging here, but [there is another `i18n_make_a_selection_text` option](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-frontend-scripts.php#L600) that we could include here. 🤷 

Something like the below...
```js
						window.alert(
							window?.wc_add_to_cart_variation_params
								?.i18n_make_a_selection_text ||
								__(
									'Please select some product options before adding this product to your cart.',
									'woocommerce-payments'
								)
						);
```

mdmoore replied:
I'm also not sure. I think this was copied over from the [original PRB implementation](https://github.com/Automattic/woocommerce-payments/blob/04b046546d34245a85fa2b29b1567b2cc25d72ac/client/payment-request/index.js#L434-L441) in @frosso tokenized cart PDP implementation and is now showing up as a commit here. What I mean is, I didn't make this change in this PR and this is [already in `develop`](https://github.com/Automattic/woocommerce-payments/blob/04b046546d34245a85fa2b29b1567b2cc25d72ac/client/tokenized-payment-request/payment-request.js#L210-L217) from the PDP tokenized cart PR. That said, may as well go ahead and update it. Thanks for pointing it out! - [c6ab890](https://github.com/Automattic/woocommerce-payments/pull/8891/commits/c6ab890ac4ccdcd74b70569424e0696db832edd0)

FangedParakeet commented on client/tokenized-payment-request/index.js:
```diff
@@ -60,17 +60,11 @@ jQuery( ( $ ) => {
+	// When the cart is updated, the PRB is removed from the page and needs to be re-initialized.
 	$( document.body ).on( 'updated_cart_totals', () => {
-		doAction( 'wcpay.payment-request.update-button-data' );
+		wooPaymentsPaymentRequest.init();
```
@frosso, in the PDP implementation, [the  `wcpay.payment-request.update-button-data` hook](https://github.com/Automattic/woocommerce-payments/blob/develop/client/tokenized-payment-request/payment-request.js#L147-L184) appears to be doing a bit more work than simply re-initialising the payment request. I _think_ this should be fine and appears to be working as expected for me, but did you previously use this hook as a placeholder or was there some greater purpose that might be lost by simply replacing it with the re-init? 🤔

frosso replied:
Thanks for flagging that - I don't think I noticed this change in my earlier review.

The hook is meant to check [if shipping requirements changed](https://github.com/Automattic/woocommerce-payments/blob/f8a38f7f68cebdc036aa720c5ae9d1d4dd2c04db/client/tokenized-payment-request/payment-request.js#L169-L169) when new products (or new options) have been added.

Doing `wooPaymentsPaymentRequest.init();`, instead, skips that step.

I could see this potentially being an issue for plugins like "Order Bump for WooCommerce", when a "simple" product is added to a cart only containing virtual products (e.g.: having the `woo-album` in the cart and adding `woo-beanie` later).
That said, I tested it locally and didn't see an issue with it. I simply copied the business logic [from the previous implementation](https://github.com/Automattic/woocommerce-payments/blob/f8a38f7f68cebdc036aa720c5ae9d1d4dd2c04db/client/payment-request/index.js#L471-L491) - I was afraid of introducing new bugs.

@mdmoore what was the reason behind this change?

mdmoore replied:
This is happening on the `updated_cart_totals` event. When the cart is updated, the PRB is removed from the page and must be re-initialized. The `wcpay.payment-request.update-button-data` hook doesn't remount the PRB. 

This is how the same event was [handled in the previous implementation](https://github.com/Automattic/woocommerce-payments/blob/f8a38f7f68cebdc036aa720c5ae9d1d4dd2c04db/client/payment-request/index.js#L665-L667). I'm guessing it was for the same reason.

frosso replied:
Thank you for the explanation, that makes sense! I didn't realize that the `updated_cart_totals` would remove the PRB from the page 🤦

FangedParakeet reviewed (commented):
Tested with GPay and the functionality here appears to be working as expected on all of the shortcode checkout, cart, and PDP pages. 🎉 

I think there are a few very minor clarifications and questions to respond to, but other than that I think this piece of the implementation looks in pretty good shape. Nice one!

frosso committed:
Merge branch 'develop' into prb-tokenized-cart-shortcode-cart-checkout

mdmoore committed:
Merge branch 'develop' into prb-tokenized-cart-shortcode-cart-checkout

mdmoore committed:
Use getPaymentRequestData() in api requests

mdmoore committed:
Update variation selection messaging

mdmoore committed:
Update tests

frosso committed:
Merge branch 'develop' into prb-tokenized-cart-shortcode-cart-checkout

FangedParakeet reviewed (approved):
I think the changes here look pretty good to me, so cheers for your reflective responses to the first round of reviews here. 

I just wanted to share something I noticed on the cart page that I encountered when updating the cart. Firstly, I noticed that if I updated the cart (e.g. increase quantity of an item), after the cart totals update, if I click the GPay button, there is some latency before the total in the GPay modal is also updated. Usually, when the modal opens the total presented will initially be the amount before the cart updated, before this updates after a few moments. Fortunately, even if I did manage to click the "Pay" button inside the GPay modal, while the incorrect total was presented, I was not allowed to complete the payment with the incorrect price, as the modal eventually refreshed with the correct total, not allowing me to continue with the incorrect total displayed. However, I'm just wondering if this latency is expected, by design, or by accident?

Also, I noticed that on each occasion when I updated the cart, when I checked the order in the admin Orders list, the Origin was always listed as "Unknown". This _only_ occurred when I completed an express checkout, while updating something in the cart during the transaction. 

I've tried to display both of this interesting quirks in this below video. 

https://github.com/Automattic/woocommerce-payments/assets/1694446/89e1085d-bb8a-4465-a1b4-66db0e683dd4

@frosso, I remember raising this "Unknown" Origin feature on one of your previous PRs and I think you resolved it by updating a meta value somewhere, but I can't remember exactly where. Do you have any idea why this might be returning on cart update or if there is some way to fix this here?

In any case, I'm still going to approve this PR now, because I think these issues are not fatal and can definitely be handled in a separate issue. So, nice one! 👍

frosso commented:
> @frosso, I remember raising this "Unknown" Origin feature on one of your previous PRs and I think you resolved it by updating a meta value somewhere, but I can't remember exactly where. Do you have any idea why this might be returning on cart update or if there is some way to fix this here?

Here is the reference to the changes: https://github.com/Automattic/woocommerce-payments/pull/8644#issuecomment-2096463754

It's weird that it's not updated correctly in this case. The `wc-order-attribution-inputs` element (used to get the order attribution values) should be present in the DOM also on shortcode cart/checkout. So I'm not sure why the order attribution is still coming out as `Unknown`.
I can investigate this separately if needed.

mdmoore committed:
Merge develop

mdmoore committed:
Fixing merge conflict

mdmoore commented:
>Firstly, I noticed that if I updated the cart (e.g. increase quantity of an item), after the cart totals update, if I click the GPay button, there is some latency before the total in the GPay modal is also updated.

>However, I'm just wondering if this latency is expected, by design, or by accident?

Good eye! The delay occurs due to the [`updateCustomer` call](https://github.com/Automattic/woocommerce-payments/blob/a63d3b90c66df54f1b0395f478e206a9f36bca44/client/tokenized-payment-request/payment-request.js#L246) that happens when the payment interface opens for the first time after initialization. Because we have to re-initialize the PRB when cart totals update, that event will fire on the first button click after totals have updated. I believe that `cartData` needs to be updated when cart totals update also, most likely via the  `wcpay.payment-request.update-button-data` hook. I quickly tried to do so, but the totals weren't updating.

>The wc-order-attribution-inputs element (used to get the order attribution values) should be present in the DOM also on shortcode cart/checkout. So I'm not sure why the order attribution is still coming out as Unknown.

Thanks for your insight here! The `wc-order-attribution-inputs` element is indeed on the cart page and values are populated initially. Because the element is hooked in via the `woocommerce_proceed_to_checkout` hook along with the PRB, it's removed from the page as well 🤦. The element is re-added without values, which explains the behavior @FangedParakeet was seeing with origin only being `unknown` on orders placed after updating cart totals. We'll just need to use a different hook on the cart page which be a little refactor, since the element is currently bundled in with the PRB wrapper.

I'll get a couple new issues created for these 2 items 👍
